src/bech32.cpp

```diff
@@ -313,69 +223,58 @@ uint32_t PolyMod(const data& v)
  * codeword, it is a multiple of G(X), so the residue is in fact just E(x) mod G(x). Note that all
  * of the (e)^j are roots of G(x) by definition, so R((e)^j) = E((e)^j).
  *
- * Syndrome returns the three values packed into a 30-bit integer, where each 10 bits is one value.
+ * Let R(x) = r1*x^5 + r2*x^4 + r3*x^3 + r4*x^2 + r5*x + r6
+ *
+ * To compute R((e)^j), we are really computing:
+ * r1*(e)^(j*5) + r2*(e)^(j*4) + r3*(e)^(j*3) + r4*(e)^(j*2) + r5*(e)^j + r6
+ *
+ * Now note that all of the (e)^(j*i) for i in [5..0] are constants and can be precomputed.
+ * But even more than that, we can consider each coefficient as a bit-string.
+ * For example, take r5 = (b_5, b_4, b_3, b_2, b_1) written out as 5 bits. Then:
+ * r5*(e)^j = b_1*(e)^j + b_2*(2*(e)^j) + b_3*(4*(e)^j) + b_4*(8*(e)^j) + b_5*(16*(e)^j)
+ * where all the (2^i*(e)^j) are constants and can be precomputed.
+ *
+ * Then we just add each of these corresponding constants to our final value based on the
+ * bit values b_i. This is exactly what is done in the Syndrome function below.
+ */
+constexpr std::array<uint32_t, 25> GenerateSyndromeConstants() {
+    std::array<uint32_t, 25> SYNDROME_CONSTS{};
+    for (int k = 1; k < 6; ++k) {
+        for (int shift = 0; shift < 5; ++shift) {
+            int16_t b = GF1024_LOG.at(1 << shift);
+            int16_t c0 = GF1024_EXP.at((997*k + b) % 1023);
+            int16_t c1 = GF1024_EXP.at((998*k + b) % 1023);
+            int16_t c2 = GF1024_EXP.at((999*k + b) % 1023);
+            uint32_t c = c2 << 20 | c1 << 10 | c0;
+            int ind = 5*(k-1) + shift;
+            SYNDROME_CONSTS[ind] = c;
+        }
+    }
+    return SYNDROME_CONSTS;
+}
+constexpr const std::array<uint32_t, 25>& SYNDROME_CONSTS = GenerateSyndromeConstants();
```

Comment:
Using a const reference here is counterproductive, it seems: the result appears to be that the *reference* is constexpr, but the value it is bound to isn't because it is a temporary, so the actual SYNDROME_CONSTS values aren't constexpr (I noticed when trying to add static_assertions on them).

Reply:
Ah okay, will revert this change then.